rush-plugins/rush-build-graph-plugin/src/GraphParser.ts

```diff
@@ -0,0 +1,197 @@
+    function getNonEmptyDependencies(node: IGraphNodeInternal): ReadonlySet<string> {
+      // If we've already computed this, return the cached result
+      let nonEmptyDependencies: Set<string> | undefined = nonEmptyDependenciesByOperation.get(node);
+      if (!nonEmptyDependencies) {
+        nonEmptyDependencies = new Set();
+        nonEmptyDependenciesByOperation.set(node, nonEmptyDependencies);
+        for (const dependencyID of node.dependencies) {
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          if (!inputNodeMap.get(dependencyID)!.command) {
+            // If the dependency is empty, recursively inherit its non-empty dependencies
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            for (const deepDependency of getNonEmptyDependencies(inputNodeMap.get(dependencyID)!)) {
+              nonEmptyDependencies.add(deepDependency);
+            }
+          } else {
+            nonEmptyDependencies.add(dependencyID);
+          }
+        }
+      }
+      return nonEmptyDependencies;
+    }
+
+    return Array.from(inputNodeMap.values())
+      .filter((node) => node.command)
+      .map((node) => {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        return { ...node, dependencies: Array.from(getNonEmptyDependencies(node)), command: node.command! };
+      });
+  }
+
+  /*
+   * Convert an operation into a graph node
+   */
+  private _operationAsHashedEntry(operation: Operation): IGraphNodeInternal {
+    const node: Partial<IGraphNodeInternal> = {
+      id: tryGetOperationId(operation),
+      task: operation.associatedPhase?.name,
+      package: operation.associatedProject?.packageName,
+      dependencies: new Set(Array.from(operation.dependencies.values(), (dep) => this.getOperationId(dep))),
+      workingDirectory: operation.associatedProject?.projectFolder,
+      // _commandToRun does not exist in the type definition of IOperationRunner,
+      //    but it's defined in subclasss ShellOperationRunner
+      command: (operation.runner as IOperationRunner & { _commandToRun: string })?._commandToRun
+    };
+
+    // command is absent because it is not required
+    const requiredFields: Array<keyof IGraphNodeInternal> = [
+      'id',
+      'task',
+      'package',
+      'dependencies',
+      'workingDirectory'
+    ];
+    const missingFields: Array<keyof IGraphNodeInternal> = requiredFields.filter((prop) => !node[prop]);
+
+    if (missingFields.length > 0) {
+      this._logger.terminal.writeErrorLine(
+        `Operation is missing required fields ${missingFields.join(', ')}: ${JSON.stringify(node)}`
+      );
+    }
+
+    // the runner is a no-op if and only if the command is empty
+    if (!!operation.runner?.isNoOp !== !node.command) {
+      throw new Error(`${node.id}: Operation runner isNoOp does not match _commandToRun existence`);
```

Comment:
I suppose I don't know when it would happen, I just thought it would be a problem if it did.. perhaps a later version of rush would break this invariant

Reply:
This could also be broken by a rush plugin tampering with the build graph. This scenario seems easy enough to handle with a warning line.

Reply:
It will happen if a plugin makes a custom runner that is not a noop.